### PR TITLE
fix condition of Reconciler triggers

### PR DESCRIPTION
**Pull request by YZ775**
- fix bug that the reconciler does not triggerd when taints of the node changes because node resource does not have `.metadata.generation`.
  - instead of using predicate.GenerationChangedPredicate, taintChangePredicate is implemented.
- remove unnecessary `MaxConcurrentReconciles` in NodeReconciler.
- add tests for the changes above.

### CODE REVIEW

File: internal/controller/utils_test.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+				{Key: "a", Value: "1", Effect: corev1.TaintEffectNoSchedule},
+				{Key: "a", Value: "1", Effect: corev1.TaintEffectPreferNoSchedule},
+				{Key: "a", Value: "2", Effect: corev1.TaintEffectNoSchedule},
+				{Key: "b", Value: "2", Effect: corev1.TaintEffectNoSchedule},
+			}
+			Expect(sorted).To(Equal(expected))
+		})
+	})
+	Context("taintsEqual function", func() {
+		It("should return true for equal taint slices", func() {
```
Comment (masa213f) [inline internal/controller/utils_test.go:37]: Please add test cases for when the taints differ.

File: internal/controller/utils_test.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+var _ = Describe("Utilities functions tests", func() {
+	Context("sortTaints function", func() {
+		It("should sort taints correctly", func() {
+			taints := []corev1.Taint{
+				{Key: "b", Value: "2", Effect: corev1.TaintEffectNoSchedule},
+				{Key: "a", Value: "2", Effect: corev1.TaintEffectNoSchedule},
+				{Key: "a", Value: "1", Effect: corev1.TaintEffectPreferNoSchedule},
+				{Key: "a", Value: "1", Effect: corev1.TaintEffectNoSchedule},
+			}
+			sorted := sortTaints(taints)
```
Comment (masa213f) [inline internal/controller/utils_test.go:18]: Please add a test case for when taints is nil.

File: internal/controller/nodetemplate_controller_test.go [hunk 3]
```diff
@@ -203,7 +203,7 @@ var _ = Describe("NodeTemplate Controller", func() {
 			L:
 				for _, node := range nodes.Items {
 					for _, taint := range node.Spec.Taints {
-						if taint.Key == "node.cybozu.io/spare" {
+						if taint.Key == SpareTaintKey {
```
Comment (masa213f) [inline internal/controller/nodetemplate_controller_test.go]: FYI:
In the case of nyallocator, the current changes (replacing direct values with a constant in tests) are no problem.
But please be aware that such changes may cause problems.

When the target program and the test reference the same constant, the test cannot detect unexpected changes to the constants. (e.g., typo)
Therefore, identifiers in tests should be written directly whenever possible (at least one).

In the case of the nyallocator, constant errors may be detected in the e2e test. It's fine as is this time.
https://github.com/cybozu-go/nyallocator/blob/v1.1.1/e2e/kind-config.yaml

File: internal/controller/utils_test.go
Comment (masa213f) [inline internal/controller/utils_test.go:1]: There are a few test cases. It is not good for work purposes.

Please implement the following as much as possible.
- Some test cases that yield different results in equivalence partitioning.
- Abnormal value cases and/or boundary value test cases.

**Review by masa213f — CHANGES_REQUESTED** [review]

File: internal/controller/nodetemplate_controller_test.go (hunk 3 above)
Comment (YZ775) [reply internal/controller/nodetemplate_controller_test.go]: I thought it's better to avoid using hard coded values because of maintenanceability. So I changed the code to reference constants.

But, it's true that we cannot detect such failures. So I fixed the test code not to using constants.

File: internal/controller/utils_test.go
Comment (YZ775) [reply internal/controller/utils_test.go:1]: Thank you. I added some test cases.

File: internal/controller/utils_test.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+			}
+			taints2 := []corev1.Taint{
+				{Key: "key1", Value: "value2", Effect: corev1.TaintEffectNoSchedule},
+			}
+			taint3 := []corev1.Taint{
+				{Key: "key2", Value: "value1", Effect: corev1.TaintEffectNoSchedule},
+			}
+			taint4 := []corev1.Taint{
+				{Key: "key1", Value: "value1", Effect: corev1.TaintEffectPreferNoSchedule},
+			}
```
Comment (masa213f) [inline internal/controller/utils_test.go]: typo?
```suggestion
			taints3 := []corev1.Taint{
				{Key: "key2", Value: "value1", Effect: corev1.TaintEffectNoSchedule},
			}
			taints4 := []corev1.Taint{
				{Key: "key1", Value: "value1", Effect: corev1.TaintEffectPreferNoSchedule},
			}
```

**Review by masa213f — CHANGES_REQUESTED** [review]

File: internal/controller/utils_test.go (hunk 4 above)
Comment (YZ775) [reply internal/controller/utils_test.go]: That's true. I fixed it.

**Review by masa213f — APPROVED** [review]
LGTM
